File: go/arena.go
package categories

type Arena struct {
    pos interface{}
    dis func(interface{}) interface{}
}

type Lens struct {
    dom, cod Arena
    observe  func(interface{}) interface{}
    interpret func(interface{}, func(interface{}) interface{}) interface{}
}

func idLens(a Arena) Lens {
    return Lens{
        dom: a, cod: a,
        observe:  func(p interface{}) interface{} { return p },
        interpret: func(p interface{}, f func(interface{}) interface{}) interface{} { return f(p) },
    }
}

func (lens23 Lens) compose(lens12 Lens) Lens {
    obs := func(p interface{}) interface{} { return lens23.observe(lens12.observe(p)) }
    int := func(p interface{}, f func(interface{}) interface{}) interface{} {
        return lens12.interpret(p, func(x interface{}) interface{} {
            return lens23.interpret(lens12.observe(p), f.(func(interface{}) interface{}))
        })
    }
    return Lens{dom: lens12.dom, cod: lens23.cod, observe: obs, interpret: int}
}

type Display interface {
    pos() interface{}
    dis(interface{}) interface{}
}

type AsFunctor struct {
    pos interface{}
    dis func(interface{}) interface{}
}

func duoidal(a1, a2, b1, b2 Arena) Lens {
    x := Pair(Combine(a1, a2), Combine(b1, b2))
    y := Combine(Pair(a1, b1), Pair(a2, b2))

    o := func(p Pos) Pos {
        p1, p2 := p.First, p.Second
        q1, q2 := p1.First, p1.Second

        pp := func(d Dis) Pos {
            return Pair(Combine(d.First, Apply(p2, d.First)), Combine(d.Second, Apply(q2, d.Second)))
        }

        return Pair(Pair(p1, q1), pp)
    }

    i := func(p Pos) func(d Dis) Dis {
        p1, p2 := p.First, p.Second
        q1, q2 := p1.First, p1.Second

        return func(d Dis) Dis {
            de1, de2 := d.First, d.Second

            return Pair(Pair(de1.First, de2.First), Pair(de1.Second, de2.Second))
        }
    }

    return Lens{o, i}
}

File: rust/src/arena.rs
use std::ops::{Deref, DerefMut};

struct Arena<Pos, Dis> {
    pos: Pos,
    dis: Box<dyn Fn(Pos) -> Dis>,
}

struct Lens<Dom, Cod> {
    observe: Box<dyn Fn(Dom) -> Cod>,
    interpret: Box<dyn Fn(Dom, Box<dyn Fn(Cod) -> Cod>) -> Box<dyn Fn(Dom) -> Dom>>,
}

impl<A> Lens<A, A> {
    fn id_lens(a: Arena<A, impl Fn(A) -> A>) -> Self {
        Lens {
            observe: Box::new(|x| x),
            interpret: Box::new(|p, f| Box::new(move |x| f(x)(x))),
        }
    }
}

impl<A1, A2, A3> Lens<A2, A3> {
    fn compose<B2, B3>(self, lens12: Lens<A1, A2>) -> Lens<A1, A3>
    where
        A2: Clone,
        A3: Clone,
        B2: 'static,
        B3: 'static,
    {
        let observe = move |p| self.observe(lens12.observe(p.clone()));
        let interpret = move |p, f| {
            lens12.interpret(
                p,
                Box::new(move |x| {
                    self.interpret(x.clone(), Box::new(move |y| f(y.clone())))
                }),
            )
        };
        Lens {
            observe: Box::new(observe),
            interpret: Box::new(interpret),
        }
    }
}

type Display<Pos, Dis> = (Pos, Box<dyn Fn(Pos) -> Dis>);
type AsFunctor<Pos, Dis, Res> = (Pos, Box<dyn Fn(Box<dyn Fn(Pos) -> Dis>) -> Res>);

struct Lens<A, B> {
    o: Box<dyn Fn(A) -> B>,
    i: Box<dyn Fn(B, Dis<B>) -> Dis<A>>,
}

impl<A, B> Lens<A, B> {
    fn duoidal(a1: Arena, a2: Arena, b1: Arena, b2: Arena) -> Lens<(Arena, Arena), (Arena, Arena)> {
        let x = (a1.combine(&a2), b1.combine(&b2));
        let y = (a1.pair(&b1), a2.pair(&b2).combine());

        let o = Box::new(
            |(p1, p2), (q1, q2)| {
                let pp = (
                    (p1, q1),
                    |d: &Dis<(Arena, Arena)>| {
                        (p2.apply(d.0.deref()), q2.apply(d.1.deref()))
                    },
                );
                pp
            },
        );

        let i = Box::new(
            |(p1, p2), (q1, q2), (de1, de2): Dis<(Arena, Arena)>| {
                (
                    (de1.0.deref().clone(), de2.0.deref().clone()),
                    (de1.1.deref().clone(), de2.1.deref().clone()),
                )
            },
        );

        Lens { o, i }
    }
}